### ISSUE TITLE
Bluetooth Connection Randomly Drops

### ISSUE THREAD

**Issue by adrian-patterson** [BasicAirData/BluetoothHelper#9]
Hi there,

First and foremost thank you for this library, it's been extremely helpful on a project using the HC-05. 
I'm not sure if this repo is still being watched or tended to, but there's an issue I've been having. I thought I might post it in case anyone might help or look to this in the future.

I use the library directly as is used in the example, and it works fantastically! I have really only this one issue and it is very minor. The way I use this library is I have it as a background service that is started in the main activity and is to run indefinitely. In the background, it continuously reads data from an HC-05 in a thread, and logs/watches the data. The issue I have is that when I exit the app (i.e. go to the home screen) for more than 2-3 minutes, and try to re-enter the app, the Bluetooth connection state somehow turns to false. The following is the output:

```
I/System.out: Receiving data
I/chatty: uid=10195(com.example.sensair) Thread-47 identical 10 lines
I/System.out: Receiving data
I/System.out: Receiving data
W/myApp: [#] Listener fired: onBluetoothHelperConnectionStateChanged = true
W/BluetoothAdapter: getBluetoothService() called with no BluetoothManagerCallback
W/BluetoothAdapter: getBluetoothService() called with no BluetoothManagerCallback
W/myApp: [#] Listener fired: onBluetoothHelperConnectionStateChanged = false
I/System.out: Disconnected
D/BluetoothSocket: close() this: android.bluetooth.BluetoothSocket@5134110, channel: 1, mSocketIS: android.net.LocalSocketImpl$SocketInputStream@965c409, mSocketOS: android.net.LocalSocketImpl$SocketOutputStream@a05660emSocket: android.net.LocalSocket@638e82f impl:android.net.LocalSocketImpl@535e3c fd:java.io.FileDescriptor@89d97c5, mSocketState: CONNECTED
W/myApp: [#] Listener fired: onBluetoothHelperConnectionStateChanged = false
I/System.out: Disconnected
W/BluetoothAdapter: getBluetoothService() called with no BluetoothManagerCallback
I/System.out: Receiving data
I/chatty: uid=10195(com.example.sensair) Thread-51 identical 7 lines
2020-11-19 16:48:48.920 12186-12366/com.example.sensair I/System.out: Receiving data
2020-11-19 16:48:48.967 12186-12368/com.example.sensair W/BluetoothAdapter: getBluetoothService() called with no BluetoothManagerCallback
W/myApp: [#] Listener fired: onBluetoothHelperConnectionStateChanged = true
W/BluetoothAdapter: getBluetoothService() called with no BluetoothManagerCallback
W/myApp: [#] Listener fired: onBluetoothHelperConnectionStateChanged = false
```

The fact that the service somehow faces a socket error when the app is reopened is very weird. I find this weird because while the app is NOT open, I can see the service reading data in the background with no problem (i.e. "Receiving Data"). But as soon as I re-open the app (the service is not stopped or restarted) and the connection disconnects.

**To Reproduce This Error:**
1. Have a service continually collecting data using this library
2. Close app and wait 2-3 minutes
3. Re-open app
4. Note in Logcat how the socket connection has somehow dropped

**Note** that this issue doesn't occur ever if I close the app and re-open it in under 2-3 minutes! So it's a very weird issue indeed.

## Edit
I am 90% sure the issue resides in the "disconnect()" function. It seems as if the socket is not properly disconnected, which results in a communications block between the phone and HC-05. If you are still interested, please reach out! I'd love to fix this with you. I will do my best to try and fix it on my own in the meanwhile.

## Edit 2
The way I fixed this was by creating a bluetooth-restarting function to reinitialize Bluetooth when there's an interruption in connectivity. Not the most elegant solution, but works for now

**Comment by GrazianoCapelli** [comment]
Dear Adrian,
I can confirm that the BluetoothHelper repo is maintained.
We are using this library as a base to implement the Bluetooth Communication Layer of our Projects.
For example the [Air Data Computer](https://github.com/BasicAirData/AirDataComputer) and its Android Interface [Air Data Bridge](https://github.com/BasicAirData/AirDataBridge).

We are not updating this library because it works good (i.e. reliably and without issues) for our use.
But we use the Library into the _Application_ Object of our apps: it works well, but it never recreates itself.

Of course, there is no Software without Bugs.

The source code of BluetoothHelper is freely available and GPL-3 licensed: if you found that the cause of your misbehaviour is in the Library Code, we would be glad to consider your suggestions, merge your Patches and publish a new Release of the Library.

Unfortunately I cannot replicate the bug you reported, and I cannot look into your code to try to find the cause of the issue you are reporting. Basing on the log you posted, it seems that some callback could be defined outside the service (and the re-creation of it, when the activity is restarted, generates the issue); or we could lose some reference into the library for some reason.
If you would post some simple code that reproduces the issue, we would be glad to make a check of it.

As a start point, I restore the right Java file on the master branch. In case We'll make every modification into the develop branch.

_As a tip: if you want to have a service that runs indefinely in background, I suggest to use a [Foreground Service](https://developer.android.com/guide/components/foreground-services) instead._

**Comment by GrazianoCapelli** [comment]
A couple of notes:

> Note that this issue doesn't occur ever if I close the app and re-open it in under 2-3 minutes! So it's a very weird issue indeed.

Here it seems that the issue could be related to your Activity and the Memory Manager

> I am 90% sure the issue resides in the "disconnect()" function. It seems as if the socket is not properly disconnected, which results in a communications block between the phone and HC-05. If you are still interested, please reach out! I'd love to fix this with you. I will do my best to try and fix it on my own in the meanwhile.

You written that the background service is to run indefinitely: why should the `disconnect()` function be called when your Activity is reopened?

**Comment by adrian-patterson** [comment]
Hi @GrazianoCapelli, thanks for taking the time to answer! This repository is dated to 2017 which is why I was unsure if it was still active, but I'm glad it is still active. You were right, the disconnect() and connect() functions were called whenever an activity was opened. Fortunately I've solved my problem, but will describe what the issue if anyone in the future wants to use this library in a service.

**Problem:** When an activity was started/stopped (onStart and onCreate), I would bind to service. When binding to the service, the service's onCreate function would be called, which would then attempt to create another connection. This made the bluetooth go haywire.
**Solution:** If you want to run a service indefinitely, make sure the connection is initialized only once, and not every time the service is bound (as I did).

So turns out it was a silly mistake on my end, so sorry for the trouble! Thanks for reaching out and helping, it is much appreciated :) Feel free to mark issue as solved or delete it altogether since the issue was on my end. Again sorry for the trouble, and thanks for the help!
